Accounting Feature

GuiMcs00 opened the pull request:
This PR implements improvements and adjustments in the Accounting project, covering dependency updates and multiple database context configurations. The changes are divided into two main parts:

1. **Dependency and Version Compatibility Adjustments**
   - Updated project packages and dependencies to compatible versions, ensuring greater stability and avoiding conflicts between libraries.
   - Synchronized package versions across services to ensure consistency, making build processes and application execution smoother and free from version compatibility issues.

2. **Multiple `DbContext` Configuration**
   - Configured multiple database contexts (`OrdersContext` and `AccountingContext`) in the `Program.cs` file, enabling the application to interact with different databases as needed.
   - Added separate connection strings in `appsettings.json` for each `DbContext`, ensuring isolation and flexibility in data management.
   - Included detailed setup instructions in the README on configuring and adding new `DbContext` instances to the Dependency Injection (DI) container, facilitating development and maintenance.

### Main Changes

- **Entities**: Created and updated entities to support new accounting data requirements.
- **Repository**: Implemented repositories for new data models, optimizing queries and CRUD operations.
- **DbContext**: Configured both `AccountingContext` and `OrdersContext` with their respective connection strings.
- **TestCases**: Added unit tests for entity validation and integration tests for repositories.
- **Dependencies**: Updated and aligned package versions to ensure compatibility.

### Purpose

These changes aim to:
- Improve project stability by eliminating dependency conflicts.
- Enable system scalability with support for multiple databases.
- Clearly document the setup process for multiple `DbContext` instances, easing future integrations and maintenance efforts.

GuiMcs00 committed:
feat add customer's email validation with FluentValidation: Implemented EmailValidator class to validate and registered in IoC Container. Inject applied in CreateCustomer endpoint.

GuiMcs00 committed:
feat(Entities, Respository, DbContext, TestCases): Implemented new features over Accounting project, including:

- Entities: Added new domain entities to ensure compatibility with business requirements.

 - Repository: Developed data access methods for new entities, providing CRUD operations.

 - DbContext: Configured `AccountingContext` to manage new entities and integrated it with PostgreSQL using Npgsql.

 - TestCases: Added integration tests for repository methods, verifying data persistence and retrieval.

GuiMcs00 committed:
chore(Dependencies, Configuration): Adjusted compatible package versions and configured multiple DbContexts

- **Dependencies**:
  - Updated project dependencies to compatible versions, ensuring stability and compatibility across all services.
  - Aligned package versions for consistency, preventing potential conflicts in runtime and build processes.

- **Configuration**:
  - Added and configured multiple `DbContext` instances (`OrdersContext`, `AccountingContext`) in `Program.cs`.
  - Implemented separate connection strings for each `DbContext` to support different databases.
  - Documented configuration steps in README to clarify setup for multi-context applications.

This commit ensures compatibility across dependencies and sets up the application for handling multiple databases efficiently.

GuiMcs00 committed:
Merge Resolve

danrleypereira commented:
> [!CAUTION]
> Pay attention to database migrations, create a unique migration per PR

danrleypereira commented on GraphOfOrders.Lib/Entities/CustomerCompany.cs:
```diff
@@ -0,0 +1,22 @@
+namespace GraphOfOrders.Lib.Entities;
+
+public class CustomerCompany
+{ 
```
add summaries in portuguese

danrleypereira commented on GraphOfOrders.Api/Program.cs:
```diff
@@ -1,24 +1,26 @@
 
 builder.Services.AddAutoMapper(typeof(MappingProfile).Assembly);
 
+// Add DbContext 
```
can we add a extension method to this ?
See AddRepoServices example.

GuiMcs00 replied:
Done ✅

GuiMcs00 committed:
feat(core): refine entity structure with abstract implementations and tenant-based initialization

- Introduced abstract implementations for entities to standardize initialization of TenantId and Id.
- Created `TenantUser` abstraction specifically for user-related entity types.
- Assigned `TenantId` to all entities participating in business rules to ensure multi-tenant support.
- Added interfaces to enforce the inclusion of abstract IDs in entities.
- Refactored enums for improved clarity and consistency across the application.

These changes improve the organization, scalability, and maintainability of the entity model while ensuring alignment with multi-tenant architecture requirements.

GuiMcs00 committed:
refactor(db): update DbContext for entity changes and improve database connection setup

- Updated `DbContext` to reflect changes in entity structure, including new abstract implementations and tenant-based initialization.
- Adjusted mappings for `TenantId` and `Id` to align with the updated entity design.
- Improved handling of `ValueConverter` for complex types like `IEnumerable<Roles>` to ensure compatibility with database storage.
- Simplified and standardized the configuration of database connections in the project.
- Enhanced support for multi-tenant architecture by ensuring proper propagation of `TenantId` in DbContext configuration.

These updates ensure seamless integration of the entity changes into the database layer and enhance the flexibility and maintainability of database connection management.

GuiMcs00 committed:
chore : improve database connection setup

GuiMcs00 committed:
chore: update package versions for compatibility and add Newtonsoft.Json

- Updated package versions to ensure compatibility with the latest framework and project dependencies.
- Added `Newtonsoft.Json` package to support advanced JSON serialization requirements.

GuiMcs00 committed:
chore(README): Updating README.md

GuiMcs00 committed:
test: improve tests and update repository to align with new entity implementation

- Refactored tests to validate changes in entity structure, including abstract implementations and tenant-based IDs.
- Enhanced test coverage for multi-tenant scenarios and complex property mappings (e.g., roles as IEnumerable).
- Updated repository methods to comply with new entity design, ensuring proper handling of TenantId and relationships.
- Simplified and streamlined test data factories for consistency and reusability.

These changes ensure that repositories and tests fully support the updated entity architecture and maintain high reliability.

GuiMcs00 committed:
feat(api): add provisional controller and DTOs for company dashboard integration

- Created a provisional controller to handle requests for company-related data needed for the dashboard.
- Introduced DTOs to structure and format company information for frontend consumption.
- Established endpoints for retrieving company details, ensuring compatibility with the frontend dashboard requirements.
- Implemented initial interaction between the backend and frontend for real-time company data display.
